src/ruby/puppetserver-lib/puppet/server/compiler.rb

```diff
@@ -9,16 +9,46 @@ def initialize
       end
 
       def compile(request_data)
-        processed_hash = convert_java_args_to_ruby(request_data)
+        catalog, logs = capture_logs do
+          processed_hash = convert_java_args_to_ruby(request_data)
 
-        node = create_node(processed_hash)
+          node = create_node(processed_hash)
 
-        catalog = Puppet::Parser::Compiler.compile(node, processed_hash['job_id'])
-        catalog.to_data_hash
+          catalog = Puppet::Parser::Compiler.compile(node, processed_hash['job_id'])
+          catalog.to_data_hash
+        end
+
+        { catalog: catalog, logs: logs }
       end
 
       private
 
+      def capture_logs(&block)
+        logs = []
+        result = nil
+        Puppet::Util::Log.close_all
+        log_dest = Puppet::Test::LogCollector.new(logs)
+        begin
+          Puppet::Util::Log.newdestination(log_dest)
+          Puppet::Util::Log.with_destination(log_dest) do
+            result = yield
+          end
```

Comment:
Is it really necessary to call both `newdestination` and `with_destination`? 

Looks like `with_destination` its self just calls `newdestination`, yields, and then calls `close` in an ensure block. Seems like we could replace the whole begin/ensure here with that.

Reply:
Oh that's nicer. I hadn't had a chance to go look at the internals of those yet, I'll give that a shot.

Reply:
Yeah, I think you could just replace the entire begin/end block with:

```
Puppet::Util::Log.with_destination(log_dest) do
  result = yield
end
```

And drop the `Puppet::Util::Log.close_all`. I think it's a good thing for logs related to this endpoint to show up in `puppetserver.log` --- just as logs from the `/puppet/v3/catalog` endpoint do.